### PR TITLE
Dev mode

### PR DESCRIPTION
**Pull request by peterhurford**
**Adds an option for `autoinstall: true` for quick development on local packages.**

```
-
  # Install a local package and re-install it every time your lockbox loads
  # regardless of package verison, to make quick development possible.
  name: yourpkg
  version: 0.1.0
  remote: local
  dir: /your/pkg
  autoinstall: true
```

### CODE REVIEW

File: R/library.R [hunk 1]
```diff
@@ -42,13 +54,24 @@ install_package <- function(locked_package, libPath, quiet) {
...
   stopifnot(is.element("dir", names(locked_package)))
 
   utils::install.packages(locked_package$dir, lib = libPath, repos = NULL
     , type = "source", quiet = quiet)
 }
 
+load_package.local <- function(locked_package) {
+  stopifnot(is.element("dir", names(locked_package)))
+  cat("Reinstalling", crayon_green(locked_package$name),
+    as.character(locked_package$version), "from", class(locked_package)[1], "\n")
```
Comment (robertzk) [inline R/library.R:70]: Maybe `if (!requireNamespace("devtools", quietly = TRUE))` stop and warn the user they must install devtools first.

File: README.md [hunk 2]
```diff
@@ -68,6 +68,14 @@ Example Lock File
   remote: local
   dir: /your/pkg
 -
+  # Install a local package and re-install it every time your lockbox loads
+  # regardless of package verison, to make quick development possible.
```
Comment (robertzk) [inline README.md]: `s/verison/version`

**Comment by robertzk** [comment]
Looks good after two minor comments are addressed.

File: R/lockbox.R [hunk 3]
```diff
@@ -49,8 +49,10 @@ lockbox.list <- function(lock, env) {
 
   ## Find the packages whose version does not match the current library.
   mismatches <- vapply(lock, version_mismatch, logical(1))
+  autoinstall_packages <- vapply(lock, function(lock) is.autoinstall_package(lock), logical(1))
+  reload_these_packages <- unlist(Map(any, mismatches, autoinstall_packages))
```
Comment (robertzk) [inline R/lockbox.R]: I would call this `load_these_packages`, since we are loading them for the first time.

File: R/lockbox.R [hunk 4]
```diff
@@ -49,8 +49,10 @@ lockbox.list <- function(lock, env) {
 
   ## Find the packages whose version does not match the current library.
   mismatches <- vapply(lock, version_mismatch, logical(1))
+  autoinstall_packages <- vapply(lock, function(lock) is.autoinstall_package(lock), logical(1))
```
Comment (robertzk) [inline R/lockbox.R]: Another way is

``` r
autoinstall_packages <- vapply(lock, is.autoinstall_package, logical(1))
```

File: R/lockbox.R (hunk 3 above)
Comment (robertzk) [inline R/lockbox.R]:
``` r
load_these_packages <- mismatches | autoinstall_packages
```

File: R/rebuild.R [hunk 5]
```diff
@@ -34,7 +34,9 @@ reset_search_path <- function() {
 }
 
 attach <- function(locked_package) {
-  if (!identical(locked_package$load, FALSE) && !locked_package$is_dependency_package) {
+  if (is.autoinstall_package(locked_package)) {
```
Comment (robertzk) [inline R/rebuild.R]: How does `load` interact with `autoinstall`? Presumably, if `load = FALSE`, we should simply re-install, but not re-load, correct?

Comment (robertzk) [inline R/rebuild.R]: In that case, this should look like:

``` r
if (!identical(locked_package$load, FALSE)) {
  if (is.autoinstall_package(locked_package)) {
    load_package(locked_package)
  } else if (!locked_package$is_dependency_package) {
    library(locked_package$name, character.only = TRUE)
  }
}
```

You could even make a `load_package.locked_package` S3 specialization be the `library()` call.

File: R/library.R (hunk 1 above)
Comment (peterhurford) [inline R/library.R:70]: done

File: README.md (hunk 2 above)
Comment (peterhurford) [inline README.md]: done

File: R/lockbox.R (hunk 3 above)
Comment (peterhurford) [inline R/lockbox.R]: done and done

File: R/lockbox.R (hunk 4 above)
Comment (peterhurford) [inline R/lockbox.R]: Done.

File: R/rebuild.R (hunk 5 above)
Comment (peterhurford) [inline R/rebuild.R]: Ok load FALSE will overwrite autoinstall TRUE now.

I can't get the last one to work because there's no autoinstall-specific class.

**Comment by robertzk** [comment]
Looks good! Let's wait for CI and merge shortly.

**Pull request merged**